open-apparel-registry/open-apparel-registry
# Create ExtendedFields

## Overview

Users can submit facility list items individually through the API or
via a list. During parsing, we search for known extended fields and create
ExtendedFields as applicable. Later, we assign a facility id to the
ExtendedFields if one has been assigned to the list item (for MATCHED
and NEW_FACILITY statuses).

For POTENTIAL_MATCH statuses, the item has not yet been assigned to a
facility. When the user confirms or rejects the match, the list item and
ExtendedFields are both linked to a facility.

Extended facility details can also be submitted via facility claim. In
this case, the facility has already been created, so the ExtendedFields
are linked to a facility immediately. A user can later update the
details supplied with their claim; in this case, changed fields are
updated, and removed fields are deleted.

Connects #1493 

## Notes

In the issue, we list `promote` as a location where a ExtendedField might need to be updated. However, while the facility is updated to point to a different FacilityListItem, the FacilityListItem (which the ExtendedField is linked to) still points to the same facility, so it doesn't seem that any changes are needed to the ExtendedField in response. 

## Testing Instructions

* Run `./scripts/manage migrate` 
* Upload a list, then process it with `./scripts/manage batch_process --list-id 16 --action parse`
* On your lists page, get the id of a successfully parsed list item, then look in the ExtendedFields in the admin. There should be five ExtendedFields containing appropriate data. They should not have a facility assigned to them yet. 
* Geocode and match the list. A new facility should be created.
```
./scripts/manage batch_process --list-id 16 --action geocode
./scripts/manage batch_process --list-id 16 --action match
```
* Get the OAR ID of the new facility, then view the ExtendedFields. They should now have the facility’s OAR ID assigned to them. (If instead of a new facility, a potential match is created, the fields will not be linked yet. Confirm the match to link the fields.)
* Create a facility using the API http://localhost:8081/api/docs/#!/facilities/facilities_create. ExtendedFields should be created and linked to the new facility (If instead of a new facility, a potential match is created, the fields will not be linked yet. Confirm the match to link the fields.)
```
{
    "country": "China",
    "name": "Nantong Jackbeanie Headwear &amp; Garment Co. Ltd.",
    "address": "No.808,the third industry park,Guoyuan Town,Nantong 226500.",
"number_of_workers": "200 - 1,000",
"native_language_name": "Jackbeanie Headwear"
}
```
* Claim a facility, then update the claim with facility details including the number of workers and the native language name. ExtendedFields should be created from the claim details. 
* Update the claim to have no number of workers listed and updating the native language name. The ExtendedField for the native language name should have been updated, and theExtendedField for the number of workers should have been deleted. 

## Checklist
